src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/repositories/FermaHaFlowRepository.java

```diff
@@ -0,0 +1,139 @@
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Ferma implementation of {@link HaFlowRepository}.
+ */
+@Slf4j
+public class FermaHaFlowRepository extends FermaGenericRepository<HaFlow, HaFlowData, HaFlowFrame>
+        implements HaFlowRepository {
+    protected final HaFlowPathRepository haFlowPathRepository;
+    protected final HaSubFlowRepository haSubFlowRepository;
+
+    public FermaHaFlowRepository(
+            FermaPersistentImplementation implementation, HaFlowPathRepository haFlowPathRepository,
+            HaSubFlowRepository haSubFlowRepository) {
+        super(implementation);
+        this.haFlowPathRepository = haFlowPathRepository;
+        this.haSubFlowRepository = haSubFlowRepository;
+    }
+
+    @Override
+    public Collection<HaFlow> findAll() {
+        return framedGraph().traverse(g -> g.V()
+                .hasLabel(HaFlowFrame.FRAME_LABEL))
+                .toListExplicit(HaFlowFrame.class).stream()
+                .map(HaFlow::new)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public boolean exists(String haFlowId) {
+        try (GraphTraversal<?, ?> traversal = framedGraph().traverse(g -> g.V()
+                .hasLabel(HaFlowFrame.FRAME_LABEL)
+                .has(HaFlowFrame.HA_FLOW_ID_PROPERTY, haFlowId))
+                .getRawTraversal()) {
+            return traversal.hasNext();
+        } catch (Exception e) {
+            throw new PersistenceException("Failed to traverse", e);
+        }
+    }
+
+    @Override
+    public Optional<HaFlow> findById(String haFlowId) {
+        return HaFlowFrame.load(framedGraph(), haFlowId).map(HaFlow::new);
+    }
+
+    @Override
+    public Optional<HaFlow> remove(String haFlowId) {
+        TransactionManager transactionManager = getTransactionManager();
+        if (transactionManager.isTxOpen()) {
+            // This implementation removes dependant entities (paths, segments, haSubFlows) in a separate transactions,
+            // so the ha-flow entity may require to be reloaded in a case of failed transaction.
+            throw new IllegalStateException("This implementation of remove requires no outside transaction");
+        }
+
+        return transactionManager.doInTransaction(() ->
+                findById(haFlowId)
+                        .map(haFlow -> {
+                            remove(haFlow);
+                            return haFlow;
+                        }));
+    }
+
+    @Override
+    protected HaFlowFrame doAdd(HaFlowData data) {
+        HaFlowFrame frame = KildaBaseVertexFrame.addNewFramedVertex(
+                framedGraph(), HaFlowFrame.FRAME_LABEL, HaFlowFrame.class);
+        HaFlow.HaFlowCloner.INSTANCE.copyWithoutSubFlowsAndPaths(data, frame);
+        frame.setSubFlows(data.getSubFlows().stream()
+                .peek(subFlow -> {
+                    if (!(subFlow.getData() instanceof HaSubFlowFrame)) {
```

Comment:
Why do you use the condition via ~not? It allows String and Integer to pass the condition.
Also, if an object in the stream is of type HaSubFlowFrame, it is not added to subFlor repo, but collected to the set later, is that intended? 
I'd create a set and then use it for add to repo and the setter.

Reply:
Thanks for explaining! 👍
I guess this is `instance of` who drew my attention,  it is sometimes tricky to get the meaning. I'd hide this checking under the super interface to get if this object is transient or not. Just for readability, something like `if(subFlow.isTransientObject()) then persist it`. But I believe this current idiom is used for all other frames? so probably not worth to touch it in just one place.

Reply:
About using not, I meant the theoretical/technical possibility. If the type is known, then, as an option, you can revert the if-clause: `if ((subFlow.getData() instanceof HaSubFlowDataImpl))`. With `not` it's like `String` is not an instance of `HaSubFlowFrame`, an absurd example, but technically true.

Reply:
Yes, We are using same instanceof in all Frame repos, so we need to change all of them for consistency